s3fs/core.py

```diff
@@ -364,6 +364,74 @@ def info(self, path, refresh=False):
             except (ClientError, ParamValidationError):
                 raise FileNotFoundError(path)
 
+    _metadata_cache = {}
+
+    def metadata(self, path, refresh=False):
+        """ Return metadata of path.
+
+        Metadata is cached unless `refresh=True`.
+
+        Parameters
+        ----------
+        path : string/bytes
+            filename to get metadata for
+        refresh : bool (=False)
+            if False, look in local cache for file metadata first
+        """
+        bucket, key = split_path(path)
+
+        if refresh or path not in self._metadata_cache:
+            response = self.s3.head_object(Bucket=bucket,
+                                           Key=key,
+                                           **self.req_kw)
+            self._metadata_cache[path] = response['Metadata']
+
+        return self._metadata_cache[path]
+
+    def getxattr(self, path, attr_name):
+        """ Get an attribute from the metadata.
+
+        Examples
+        --------
+        >>> mys3fs.getxattr('mykey', 'attribute_1')  # doctest: +SKIP
+        'value_1'
+        """
+        xattr = self.metadata(path)
+        if attr_name in xattr:
+            return xattr[attr_name]
+        return None
+
+    def setxattr(self, path, **kw_args):
+        """ Set metadata.
+
+        Attributes have to be of the form documented in the `Metadata Reference`_.
+
+        Examples
+        --------
+        >>> mys3file.setxattr(attribute_1='value1', attribute_2='value2')  # doctest: +SKIP
+
+        .. Metadata Reference:
+        http://docs.aws.amazon.com/AmazonS3/latest/dev/UsingMetadata.html#object-metadata
+        """
+
+        bucket, key = split_path(path)
+        metadata = self.metadata(path)
+        metadata.update(**kw_args)
+
+        # remove all keys that are None
+        for kw_key in kw_args:
+            if kw_args[kw_key] is None:
+                metadata.pop(kw_key, None)
+
+        self.s3.copy_object(CopySource="{}/{}".format(bucket, key),
+                            Bucket=bucket,
+                            Key=key,
+                            Metadata=metadata,
+                            MetadataDirective='REPLACE')
+
+        # refresh metadata
+        self.metadata(path, refresh=True)
```

Comment:
If the operation succeeded, can we not assume that the metadata was correctly updated, rather than refreshing?